### ISSUE TITLE
[BUG] FLOAT32 rounding more inaccurate than necessary

### ISSUE THREAD

**Issue by jlowe** [rapidsai/cudf#14528]
**Describe the bug**
When rounding single-precision floats to a specified number of decimal places, the result can be slightly inaccurate due to the intermediate computations being forced into FLOAT32 as well.  round.cu has rounding functors for non-fixed-point types, but all of the intermediate results are in the same type as the input rather than the highest precision type, double.  This means more error is introduced during the rounding computation than is necessary.

**Steps/Code to reproduce bug**
The following code demonstrates the problem:
```c++
#include <cudf/round.hpp>
#include <cudf_test/column_utilities.hpp>
#include <cudf_test/column_wrapper.hpp>
#include <cudf_test/debug_utilities.hpp>

int main(int argc, char** argv) {
  auto const input =
    cudf::test::fixed_width_column_wrapper<float>{6.121944898040965e-05f};
  cudf::test::print(input);
  auto const result = cudf::round(input, 10, cudf::rounding_method::HALF_UP);
  cudf::test::print(*result);
  return 0;
}
```

Rounding the value to the tenth decimal place should round _down_ to approximately 6.12194e-05 but instead the value is rounded _up_ to approximately 6.12195e-05 as shown in the output when running the program:
```
6.1219449e-05
6.12194999e-05
```

**Expected behavior**
FLOAT32 rounding should use FLOAT64 for intermediate results during computation to try to avoid injecting errors beyond what is necessary when dealing with floating point numbers.  When I manually performed the computations on this example input value for round.cu's half_up_positive logic but using double instead of float for the intermediate values, the answer came out rounded down as expected.

It seems that the functors for floating point rounding in round.cu should _not_ be using whatever the input type is but rather `double` explicitly to avoid unnecessary additional error during the computation.

**Comment by bdice** [comment]
Looking at the implementation of our rounding code (specifically [half_up_positive](https://github.com/rapidsai/cudf/blob/e15290a373ff0c84c85c2c0e940e69377a66cf96/cpp/src/round/round.cu#L103-L112)), I am concerned that it's not just a matter of `float` becoming `double`. I suspect `double` values could see the same type of problem, if the number of decimal places is large enough. I think we may need to use an algorithm that is more numerically stable and has less precision loss due to truncation. See also #14169, which is a separate problem with rounding algorithms in libcudf affecting float-to-decimal conversions.

**Comment by jlowe** [comment]
Yes, totally agree that double will have similar issues with larger decimal place rounding.  As such I don't see this so much as "solving" all the rounding issues as significantly improving the results for FLOAT32 with minimal effort.

**Comment by wence-** [comment]
That said, I claim that `half_up` rounding doesn't make much sense as tiebreaking scheme for binary floating point values since the only fractional floating point values that exactly end in 5 are those of the form $2^{-n}$ for $n \ge 1$.

**Comment by pmattione-nvidia** [comment]
I started looking into this, and noticed a problem with our current implementation: HALF_UP rounding is bugged for negative numbers: 

if you round (e.g.) **-0.5f** to zero decimal places, it should round **up** to zero, but instead results in **-1** because it calls roundf(), which rounds **away** from zero. 

Do we want to change the rounding code to round up, or change the name/comment to ROUND_AWAY?

**Comment by pmattione-nvidia** [comment]
That's fine with me, I'll fix the code comment then in round.hpp (which should instead point to wikipedia's rounding half away from zero).

**Comment by harrism** [comment]
> When rounding single-precision floats to a specified number of decimal places

The above is a nonsense statement, in my book. By now you should know my opinion on this: libcudf should not support `round` to decimal places on binary floating-point types.

https://github.com/rapidsai/cudf/issues/406
https://github.com/rapidsai/cudf/issues/1270
https://github.com/rapidsai/cudf/issues/1340
https://github.com/rapidsai/cudf/issues/7195

I'm sure there are more.

**Comment by pmattione-nvidia** [comment]
The most accurate way to round would be to convert floating-point to fixed_point (to more decimal places than float/doubles precision), round as desired, then convert that back to floating. 

But as @harrism brings up this will still produce "unexpected" results (e.g.  1.151f to 2 decimal places -> fixed point 115 w/ scale -1 -> 1.1499999f -> string "1.14").  

So I think the question is: do we want to do these conversions in this kernel, or should we block using floating point for the round kernel, and instead require users to explicitly choose the chain of operations above?

**Comment by harrism** [comment]
I think licudf shouldn't support rounding binary float types to specified number of decimal places. Just like the C++ standard provides no function for this. 

https://en.cppreference.com/w/cpp/numeric/math/round 
> Computes the nearest integer value to num (in floating-point format)
